### PR TITLE
feat: add --metadata-include and --metadata-exclude parameters to unstructured-ingest

### PR DESCRIPTION
**Pull request by natygyoon**
Currently, the default behavior of [unstructured-ingest](https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/ingest/main.py) is to [drop the metadata field](https://github.com/Unstructured-IO/unstructured/blob/70420b5/unstructured/ingest/interfaces.py#L97) “filename” from the json output.

However, it would be better if the user could specify what metadata fields to drop (if they exist) or what fields to include (drop all other fields).

### CODE REVIEW

File: unstructured/ingest/main.py [hunk 1]
```diff
@@ -103,6 +133,20 @@ def run(self):
 
 
 @click.command()
+@click.option(
+    "--metadata-include",
+    cls=MutuallyExclusiveOption,
+    default="",
+    help="If set, include the specified metadata fields if they exist and drop all other fields.",
```
Comment (ryannikolaidis) [inline unstructured/ingest/main.py:140]: How does one assign this list when calling? Is it a single string with comma separated values? or do you make multiple `--metadata-include X --metadata-include Y` args? Maybe helpful to have an example in the help. (Same for exclude)

File: unstructured/ingest/interfaces.py [hunk 2]
```diff
@@ -47,6 +47,8 @@ class BaseConnectorConfig(ABC):
     # where to write structured data outputs
     output_dir: str
     re_download: bool = False
+    metadata_include: str = ""
+    metadata_exclude: str = ""
```
Comment (ryannikolaidis) [inline unstructured/ingest/interfaces.py:51]: This will mean that the mutual exclusivity is only actually defined at the click() interface level. By testing `if metadata_exclude` first in the BaseIngestDoc, I think your logic still works, but I wonder if it'd be helpful to be more explicit here? Make these say Optional(str) with default None to distinguish between the user specifying an empty string versus not setting the value. You could then also validate exclusivity there by actually testing `is None`. Thoughts?

**Comment by ryannikolaidis** [comment]
This is looking really good! Just dropped a few questions.

File: unstructured/ingest/interfaces.py (hunk 2 above)
Comment (natygyoon) [reply unstructured/ingest/interfaces.py]: Yeah, that's a really good idea to make them optional string and more explicitly signal the user that at most one of them should be set. Just pushed a fix.

Comment (ryannikolaidis) [reply unstructured/ingest/interfaces.py]: awesome! taking a look now.

File: unstructured/ingest/main.py [hunk 3]
```diff
@@ -35,6 +36,35 @@
...
+
+    def handle_parse_result(self, ctx, opts, args):
+        if self.mutually_exclusive.intersection(opts) and self.name in opts:
+            raise UsageError(
+                "Illegal usage: `{}` is mutually exclusive with "
+                "arguments `{}`.".format(
+                    self.name,
+                    ", ".join(self.mutually_exclusive),
+                ),
+            )
```
Comment (ryannikolaidis) [inline unstructured/ingest/main.py:59]: I just tried:
```
python3 -m unstructured.ingest.main --gitlab-url https://gitlab.com/gitlab-com/content-sites/docsy-gitlab --git-file-glob '*.md,*.txt' --reprocess --metadata-exclude page_number --metadata-include page_number
```

This succeeded. I thought I should have seen this error message, right?

File: unstructured/ingest/interfaces.py [hunk 4]
```diff
@@ -94,7 +99,16 @@ def process_file(self):
         self.isd_elems_no_filename = []
         for elem in isd_elems:
             # type: ignore
-            elem["metadata"].pop("filename", None)  # type: ignore[attr-defined]
+            if self.config.metadata_exclude is not None:
+                ex_list = self.config.metadata_exclude.split(",")
+                for ex in ex_list:
+                    elem["metadata"].pop(ex, None)  # type: ignore[attr-defined]
+            elif self.config.metadata_include is not None:
```
[the remark this answers is not in the document]
Comment (natygyoon) [reply unstructured/ingest/interfaces.py:106]: I agree that adding additional check here ensures the mutual exclusiveness even though we are checking it on a higher level. Just pushed the fix.

**Review by ryannikolaidis — APPROVED** [review]
Looks good, nice work!
